File: make_DTI_lists.m
function make_DTI_lists(exptDir,subIDs,outName,nScans,idStr)
%MAKE_DTI_LISTS - Setup Pipeline lists for DTI preprocessing workflow
%For a given experiment directory, this script checks to see how many DTI
%scan repetitions are present, and generates the appropriate input lists
%for the DTI_preprocessing.pipe workflow in the <exptDir>/PIPELINE directory.
%
% Syntax: make_DTI_lists(exptDir,subDIs,outName,nScans,idStr)
%
% Inputs:
%    exptDir - Path to base experiment directory. Should contain SUBJECTS and
%              PIPELINE directories
%    subIDs  - Vector of subject ID folders in <exptDir>/SUBJECTS (for now, must
%              be numeric)
%    outName - Desired output directory name that will be created in each
%              <exptDir>/SUBJECTS/<subID>/ directory
%    nScans  - Choose the number of scans to average
%    idStr   - Specify a string to identify raw DWI series
%
% Outputs:
%
% Output files:
%
% Example:
%    exptDir = '/path/to/exptDir/';
%    outName = '2avg';
%    nScans  = 2;
%    idStr   = '30DIR';
%    D       = dir(fullfile(exptDir, 'SUBJECTS/20*'));
%    subIDs  = str2num(vertcat(D.name));
%    %subIDs  = 20037;                              % Also can specify subjects(s)
%    %load(fullfile(exptDir, 'SCRIPTS/subIDs.txt')) % or load up a list
%    make_DTI_lists(exptDir,subIDs,outName,nScans,idStr)
%
% Notes:
% 1. To clear out a whole area, do something like: rm SUBJECTS/*/1avg/dtifit/*
% 2. Change 'SUBJECTS' to 'TEST' at line 47 to use a test area
%
% Other m-files required: none
% Subfunctions: none
% MAT-files required: none
%
% See also: DTI_preprocessing.pipe

% Author: John Colby (johncolby@ucla.edu)
% UCLA Developmental Cognitive Neuroimaging Group (Sowell Lab)
% Jan 2010

%% Check input arguments
if nargin < 4 || isempty(nScans) || isempty(outName) || isempty(exptDir)
    error('Must specify input arguments')
end

%% Open files for output
inpt = fopen(fullfile(exptDir, 'PIPELINE', 'inpt.list'),'wt');
dti  = fopen(fullfile(exptDir, 'PIPELINE', 'dti.list'),'wt');
data = fopen(fullfile(exptDir, 'PIPELINE', 'data.list'),'wt');
bet  = fopen(fullfile(exptDir, 'PIPELINE', 'bet.list'),'wt');
mask = fopen(fullfile(exptDir, 'PIPELINE', 'mask.list'),'wt');
bvec = fopen(fullfile(exptDir, 'PIPELINE', 'bvec.list'),'wt');
bval = fopen(fullfile(exptDir, 'PIPELINE', 'bval.list'),'wt');
dtk  = fopen(fullfile(exptDir, 'PIPELINE', 'dtk.list'),'wt');
dtk2 = fopen(fullfile(exptDir, 'PIPELINE', 'dtk2.list'),'wt');
fa   = fopen(fullfile(exptDir, 'PIPELINE', 'fa.list'),'wt');

%% Generate Pipeline input lists
for i=1:length(subIDs) % Loop over subject ID
    subStr = num2str(subIDs(i));
    
    % Change 'SUBJECTS' to 'TEST' to use a test area
    subDir = fullfile(exptDir, 'SUBJECTS', subStr);
    analysisDir = fullfile(subDir, outName);
    
    % Check to see if there are DTI scans present, and how many
    D = dir(fullfile(subDir, 'RAW', sprintf('*%s*.nii.gz', idStr)));
    
    if length(D)>=nScans
        % Make folders if needed
        if ~isdir(fullfile(analysisDir, 'dtifit')), mkdir(fullfile(analysisDir, 'dtifit')); end
        if ~isdir(fullfile(analysisDir, 'track')), mkdir(fullfile(analysisDir, 'track')); end
        if ~isdir(fullfile(analysisDir, 'diffusion_toolkit')), mkdir(fullfile(analysisDir, 'diffusion_toolkit')); end
        
        % Only do tensor fitting for subjects who don't have it already
        if isempty(dir(fullfile(analysisDir, 'dtifit/dti*'))) || isempty(dir(fullfile(analysisDir, 'diffusion_toolkit/dti*')))
            [tmp files] = unix(sprintf('echo `ls %s | head -%d`', fullfile(subDir, sprintf('*%s*.nii.gz', idStr)), nScans));
            fprintf(inpt, sprintf('%s', files));
            fprintf(dti,  sprintf('%s\n', fullfile(analysisDir, 'dtifit')));
            fprintf(data, sprintf('%s\n', fullfile(analysisDir, 'track/data.nii.gz')));
            fprintf(bet,  sprintf('%s\n', fullfile(analysisDir, 'track/nodif_brain.nii.gz')));
            fprintf(mask, sprintf('%s\n', fullfile(analysisDir, 'track/nodif_brain_mask.nii.gz')));
            fprintf(bvec, sprintf('%s\n', fullfile(exptDir, 'PIPELINE/grad', sprintf('bvecs%d', nScans))));
            fprintf(bval, sprintf('%s\n', fullfile(exptDir, 'PIPELINE/grad', sprintf('bvals%d', nScans))));
            fprintf(dtk,  sprintf('%s\n', fullfile(analysisDir, 'diffusion_toolkit', 'dti')));
            fprintf(dtk2, sprintf('%s\n', fullfile(analysisDir, 'diffusion_toolkit', 'dti.trk')));
            fprintf(fa,   sprintf('%s\n', fullfile(analysisDir, 'diffusion_toolkit', 'dti_fa.nii.gz')));
            
            % Make links to the correct bvals/bvecs in the track folder
            if ~exist(fullfile(analysisDir, 'track/bvals'), 'file')
                unix(sprintf('ln -s %s %s', fullfile(exptDir, 'PIPELINE/grad', sprintf('bvals%d', nScans)), ...
                    fullfile(analysisDir, 'track/bvals')));
                unix(sprintf('ln -s %s %s', fullfile(exptDir, 'PIPELINE/grad', sprintf('bvecs%d', nScans)), ...
                    fullfile(analysisDir, 'track/bvecs')));
            end
        end
    end
end

%% Clean up
fclose all;

% Set permissions so Pipeline can read the files
unix(sprintf('chmod a+rw %s', fullfile(exptDir, 'PIPELINE/*.list')));

File: make_DTI_lists.py
#!/usr/bin/env python
import argparse

parser = argparse.ArgumentParser(formatter_class=argparse.RawDescriptionHelpFormatter,
description='\
For a given experiment directory, this script checks to see how many DTI scan\n\
repetitions are present, and generates the appropriate input lists for the\n\
DTI_preprocessing.pipe workflow in the <exptDir>/PIPELINE directory.',
epilog='Example: make_DTI_lists.py /path/to/exptDir 2avg 2 30DIR -s `ls /path/to/exptDir/SUBJECTS | grep <some pattern goes here>`')

parser.add_argument('exptDir', action='store',
                    help='Path to base experiment directory. Should contain SUBJECTS and PIPELINE directories.')
parser.add_argument('outName', action='store',
                    help='Desired output directory name that will be created in each <exptDir>/SUBJECTS/<subID>/ directory.')
parser.add_argument('nScans', action='store', type=int,
                    help='Choose the number of scans to average.')
parser.add_argument('idStr', action='store',
                    help='Specify a string to identify raw DWI series.')
parser.add_argument('-s', action='store', nargs='+', dest='subIDs',
                    help='Subset of subject ID directories in <exptDir>/SUBJECTS to process. Default is to process all subjects.')

args = parser.parse_args()

import os
from os.path import *
from glob import *

def makeDirs(analysisDir):
    # Setup output directory structure if needed
    
    if not isdir(join(analysisDir, 'dtifit')):
        os.makedirs(join(analysisDir, 'dtifit'))
    if not isdir(join(analysisDir, 'track')):
        os.makedirs(join(analysisDir, 'track'))
    if not isdir(join(analysisDir, 'diffusion_toolkit')):
        os.makedirs(join(analysisDir, 'diffusion_toolkit'))

def make_DTI_lists(exptDir, outName, nScans, idStr, subIDs=''):
    # Function to generate Pipeline lists
    
    # Convert exptDir to full path
    exptDir = abspath(exptDir)
    
    # Get full subject list if subIDs is not given
    if not subIDs:
        subIDs = sorted(os.listdir(join(exptDir, 'SUBJECTS')))
    
    # Open files for output
    with open(join(exptDir, 'PIPELINE', 'inpt.list'), 'w') as inpt,\
         open(join(exptDir, 'PIPELINE', 'dti.list'), 'w')  as dti,\
         open(join(exptDir, 'PIPELINE', 'data.list'), 'w') as data,\
         open(join(exptDir, 'PIPELINE', 'bet.list'), 'w')  as bet,\
         open(join(exptDir, 'PIPELINE', 'mask.list'), 'w') as mask,\
         open(join(exptDir, 'PIPELINE', 'bvec.list'), 'w') as bvec,\
         open(join(exptDir, 'PIPELINE', 'bval.list'), 'w') as bval,\
         open(join(exptDir, 'PIPELINE', 'dtk.list'), 'w')  as dtk,\
         open(join(exptDir, 'PIPELINE', 'dtk2.list'), 'w') as dtk2,\
         open(join(exptDir, 'PIPELINE', 'fa.list'), 'w')   as fa:
         
        # Loop over subject ID
        for subject in subIDs:
            
            # Change 'SUBJECTS' to 'TEST' to use a test area
            subDir      = join(exptDir, 'SUBJECTS', subject)
            analysisDir = join(subDir, outName)
            
            # Check to see if there are DTI scans present, and how many
            files = sorted(glob(join(subDir, 'RAW', '*' + idStr + '*.nii.gz')))
            
            print(str(subject) + '\t'),
            
            if len(files) >= nScans:
                makeDirs(analysisDir)
                
                # Only set up tensor fitting to run for subjects who don't have
                # files already
                if not (glob(join(analysisDir, 'dtifit/dti*')) or\
                        glob(join(analysisDir, 'diffusion_toolkit/dti*'))):
                    print >> inpt, ' '.join(files[0:nScans])
                    print >> dti,  join(analysisDir, 'dtifit')
                    print >> data, join(analysisDir, 'track/data.nii.gz')
                    print >> bet,  join(analysisDir, 'track/nodif_brain.nii.gz')
                    print >> mask, join(analysisDir, 'track/nodif_brain_mask.nii.gz')
                    print >> bvec, join(exptDir, 'PIPELINE/grad', 'bvecs' + str(nScans))
                    print >> bval, join(exptDir, 'PIPELINE/grad', 'bvals' + str(nScans))
                    print >> dtk,  join(analysisDir, 'diffusion_toolkit', 'dti')
                    print >> dtk2, join(analysisDir, 'diffusion_toolkit', 'dti.trk')
                    print >> fa,   join(analysisDir, 'diffusion_toolkit', 'dti_fa.nii.gz')
                    
                    print('OK\t'),
                    
                    # Make links to the correct bvals/bvecs in the track folder
                    # for bedpostx
                    try:
                        os.symlink(join(exptDir, 'PIPELINE/grad', 'bvals' + str(nScans)),\
                                   join(analysisDir, 'track/bvals'))
                    except OSError as err:
                        print('bvals: ' + str(err) + '\t'),
                    try:
                        os.symlink(join(exptDir, 'PIPELINE/grad', 'bvecs' + str(nScans)),\
                                   join(analysisDir, 'track/bvecs'))
                    except OSError as err:
                        print('bvecs: ' + str(err))
                else:
                    print('DTI files already present')
            else:
                print('Not enough scans')

make_DTI_lists(args.exptDir, args.outName, args.nScans, args.idStr, args.subIDs)